drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -0,0 +1,2897 @@
+if (TD->HwTD.Buffer)
+    DPRINT("Buffer          - %p\n", TD->HwTD.Buffer);
+
+if (TD->SetupPacket.bmRequestType.B)
+    DPRINT("bmRequestType   - %02X\n", TD->SetupPacket.bmRequestType.B);
+if (TD->SetupPacket.bRequest)
+    DPRINT("bRequest        - %02X\n", TD->SetupPacket.bRequest);
+if (TD->SetupPacket.wValue.W)
+    DPRINT("wValue          - %04X\n", TD->SetupPacket.wValue.W);
+if (TD->SetupPacket.wIndex.W)
+    DPRINT("wIndex          - %04X\n", TD->SetupPacket.wIndex.W);
+if (TD->SetupPacket.wLength)
+    DPRINT("wLength         - %04X\n", TD->SetupPacket.wLength);
+
+    DPRINT("PhysicalAddress - %p\n", TD->PhysicalAddress);
+    DPRINT("Flags           - %X\n", TD->Flags);
+    DPRINT("NextHcdTD       - %p\n", TD->NextHcdTD);
+    DPRINT("UhciTransfer    - %p\n", TD->UhciTransfer);
+}
+
+VOID
+NTAPI
+UhciFixDataToggle(IN PUHCI_EXTENSION UhciExtension,
+                  IN PUHCI_ENDPOINT UhciEndpoint,
+                  IN PUHCI_HCD_TD TD,
+                  IN BOOL DataToggle)
+{
+    DPRINT_UHCI("UhciFixDataToggle: UhciExtension - %p, UhciEndpoint - %p, DataToggle - %X\n",
+                UhciExtension,
+                UhciEndpoint,
+                DataToggle);
+
+    while (TD)
+    {
+        TD->HwTD.Token.DataToggle = (TD->HwTD.Token.DataToggle == FALSE);
+        DataToggle = (DataToggle == FALSE);
+
+        TD = TD->NextHcdTD;
+    }
+
+    UhciEndpoint->DataToggle = DataToggle;
+}
+
+VOID
+NTAPI
+UhciCleanupFrameListEntry(IN PUHCI_EXTENSION UhciExtension,
+                          IN ULONG Index)
+{
+    PUHCI_HC_RESOURCES UhciResources;
+    ULONG_PTR PhysicalAddress;
+    ULONG HeadIdx;
+
+    UhciResources = UhciExtension->HcResourcesVA;
+
+    if (Index == 0)
+    {
+        PhysicalAddress = UhciExtension->StaticTD->PhysicalAddress;
+
+        UhciResources->FrameList[0] = PhysicalAddress |
+                                      UHCI_FRAME_LIST_POINTER_TD;
+    }
+    else
+    {
+        HeadIdx = (INTERRUPT_ENDPOINTs - ENDPOINT_INTERRUPT_32ms) +
+                  (Index & (ENDPOINT_INTERRUPT_32ms - 1));
+
+        PhysicalAddress = UhciExtension->IntQH[HeadIdx]->PhysicalAddress;
+
+        UhciResources->FrameList[Index] = PhysicalAddress |
+                                          UHCI_FRAME_LIST_POINTER_QH;
+    }
+}
+
+VOID
+NTAPI 
+UhciCleanupFrameList(IN PUHCI_EXTENSION UhciExtension,
+                     IN BOOLEAN IsAllEntries)
+{
+    ULONG NewFrameNumber;
+    ULONG OldFrameNumber;
+    ULONG ix;
+
+    DPRINT_UHCI("UhciCleanupFrameList: [%p] All - %x\n",
+                UhciExtension, IsAllEntries);
+
+    if (InterlockedIncrement(&UhciExtension->LockFrameList) != 1)
+    {
+        InterlockedDecrement(&UhciExtension->LockFrameList);
+        return;
+    }
+
+    NewFrameNumber = UhciGet32BitFrameNumber(UhciExtension);
+    OldFrameNumber = UhciExtension->FrameNumber;
+
+    if ((NewFrameNumber - OldFrameNumber) < UHCI_FRAME_LIST_MAX_ENTRIES &&
```

Comment:
Do frame numbers roll over or are they monotonously increasing? If the former, you may get an integer overflow here; if the latter, `ASSERT(NewFrameNumber >= OldFrameNumber)` above

Reply:
Yes, in theory there may be an overflow here. Since MAX_ULONG == 0xFFFFFFFF == 4,294,967,295, and in NewFrameNumber are stored milliseconds (1 USB frame == 1 millisecond), then dividing by 1000 we get ~ 4,294,967 seconds. Dividing by (60 * 60 * 24), we get ~ 50 days. Thus, approximately 50 days after switching on the computer, we can lose isochronous transmissions (usually an audio signal).